src/deluge/gui/ui/keyboard/keyboard_screen.cpp

```diff
@@ -0,0 +1,686 @@
+	if (soundEditorResult != ActionResult::NOT_DEALT_WITH) {
+		return soundEditorResult;
+	}
+
+	// Exit if pad is enabled but UI in wrong mode, this was removed as it prevented from changing root note
+	// if (!isUIModeWithinRange(padActionUIModes)
+	//     && velocity) {
+	// 	return ActionResult::DEALT_WITH;
+	// }
+
+	// Pad pressed down, add to list if not full
+	if (velocity) {
+		int freeSlotIdx = -1;
+		for (int idx = 0; idx < MAX_NUM_KEYBOARD_PAD_PRESSES; ++idx) {
+			// Free slot found
+			if (!pressedPads[idx].active) {
+				freeSlotIdx = idx;
+				continue;
+			}
+
+			// Pad was already active
+			if (pressedPads[idx].active && pressedPads[idx].x == x && pressedPads[idx].y == y) {
+				freeSlotIdx = -1; // If a free slot was found previously, reset it so we don't write a second entry
+				break;
+			}
+		}
+
+		// Store active press in the free slot
+		if (freeSlotIdx != -1) {
+			pressedPads[freeSlotIdx].x = x;
+			pressedPads[freeSlotIdx].y = y;
+			pressedPads[freeSlotIdx].active = true;
+		}
+	}
+
+	// Pad released, remove from list
+	else {
+		for (int idx = 0; idx < MAX_NUM_KEYBOARD_PAD_PRESSES; ++idx) {
+			// Pad was already active
+			if (pressedPads[idx].active && pressedPads[idx].x == x && pressedPads[idx].y == y) {
+				pressedPads[idx].active = false;
+				break;
+			}
+		}
+	}
+
+	evaluateActiveNotes();
+
+	// Handle setting root note
+	if (currentUIMode == UI_MODE_SCALE_MODE_BUTTON_PRESSED) {
+		if (sdRoutineLock) {
+			return ActionResult::REMIND_ME_OUTSIDE_CARD_ROUTINE;
+		}
+
+		// We probably couldn't have got this far if it was a Kit, but let's just check
+		if (currentSong->currentClip->output->type != InstrumentType::KIT && lastNotesState.count == 0
+		    && currentNotesState.count == 1) {
+			exitScaleModeOnButtonRelease = false;
+			if (getCurrentClip()->inScaleMode) {
+				instrumentClipView.setupChangingOfRootNote(currentNotesState.notes[0].note);
+				requestRendering();
+				displayCurrentScaleName();
+			}
+			else {
+				enterScaleMode(currentNotesState.notes[0].note);
+			}
+		}
+	}
+	else {
+		updateActiveNotes();
+	}
+
+	requestRendering();
+	return ActionResult::DEALT_WITH;
+}
+
+void KeyboardScreen::evaluateActiveNotes() {
+	lastNotesState = currentNotesState;
+	layoutList[getCurrentClip()->keyboardState.currentLayout]->evaluatePads(pressedPads);
+	currentNotesState = *layoutList[getCurrentClip()->keyboardState.currentLayout]->getNotesState();
+}
+
+void KeyboardScreen::updateActiveNotes() {
+	Instrument* activeInstrument = getActiveInstrument();
+	char modelStackMemory[MODEL_STACK_MAX_SIZE];
+	ModelStack* modelStack = setupModelStackWithSong(modelStackMemory, currentSong);
+	bool clipIsActiveOnInstrument = makeCurrentClipActiveOnInstrumentIfPossible(modelStack);
+
+	// Handle added notes
+	for (uint8_t idx = 0; idx < currentNotesState.count; ++idx) {
+		uint8_t newNote = currentNotesState.notes[idx].note;
+		if (lastNotesState.noteEnabled(newNote)) {
+			continue; // Note was enabled before
```

Comment:
The original code contained some comments and extra logic to prevent stuck notes. I liked this type of abstraction since layouts don't need to care about sounding at all and the KeyboardLayout class only accesses external objects to obtain information, it has no side effects or changes external state. 

It also means layouts can easily implement arpeggiators or chord/harmony generators where they only need to care about the notes and while I'm pretty sure it doesn't make a difference the auditioning of the notes happens pretty tightly which should prevent any delays between them

Reply:
Layouts that do more complex sequencing is a good point. LGTM.